ledger/sandbox-on-x/src/main/scala/com/daml/ledger/sandbox/ConflictCheckingLedgerBridge.scala

```diff
@@ -0,0 +1,595 @@
+import com.daml.ledger.configuration.{
+  Configuration,
+  LedgerId,
+  LedgerInitialConditions,
+  LedgerTimeModel,
+}
+import com.daml.ledger.offset.Offset
+import com.daml.ledger.participant.state.index.v2.IndexService
+import com.daml.ledger.participant.state.v2.Update.CommandRejected.FinalReason
+import com.daml.ledger.participant.state.v2._
+import com.daml.ledger.sandbox.ConflictCheckingLedgerBridge.Submission._
+import com.daml.ledger.sandbox.SoxRejection.{
+  CausalMonotonicityViolation,
+  DuplicateKey,
+  GenericRejectionFailure,
+  InconsistentContractKey,
+  UnknownContracts,
+}
+import com.daml.lf.data.Time.Timestamp
+import com.daml.lf.data.{Ref, Time}
+import com.daml.lf.engine.Blinding
+import com.daml.lf.transaction.Transaction.{KeyActive, KeyCreate, NegativeKeyLookup}
+import com.daml.lf.transaction.{BlindingInfo, CommittedTransaction, GlobalKey, SubmittedTransaction}
+import com.daml.logging.{ContextualizedLogger, LoggingContext}
+import com.daml.metrics.{InstrumentedSource, Metrics, Timed}
+import com.daml.platform.apiserver.execution.MissingContracts
+import com.daml.platform.store.appendonlydao.events._
+import com.daml.telemetry.TelemetryContext
+import com.google.common.primitives.Longs
+import com.google.rpc.code.Code
+import com.google.rpc.status.Status
+import com.daml.lf.transaction.Transaction.{
+  KeyInput => TxKeyInput,
+  KeyInputError => TxKeyInputError,
+}
+
+import java.util.UUID
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+import java.util.concurrent.{CompletableFuture, CompletionStage}
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.{Failure, Success, Try}
+
+case class ConflictCheckingLedgerBridge(
+    participantId: Ref.ParticipantId,
+    ledgerId: LedgerId,
+    maxDedupSeconds: Int,
+    submissionBufferSize: Int,
+    // Hack needed for avoiding the cyclic dependency in the kvutils app Runner
+    indexServiceRef: AtomicReference[Option[IndexService]],
+    metrics: Metrics,
+)(implicit
+    mat: Materializer,
+    loggingContext: LoggingContext,
+    servicesExecutionContext: ExecutionContext,
+) extends ReadService
+    with WriteService
+    with AutoCloseable {
+  private val offsetIdx = new AtomicLong(0L)
+  private lazy val indexService =
+    indexServiceRef.get().getOrElse(throw new RuntimeException("ContractStore not there yet"))
+
+  private var stateUpdatesWasCalledAlready = false
+
+  import ConflictCheckingLedgerBridge._
+
+  private[this] val logger = ContextualizedLogger.get(getClass)
+
+  private type KeyInputs = Either[TxKeyInputError, Map[Key, TxKeyInput]]
+
+  private val (conflictCheckingQueue, source) =
+    InstrumentedSource
+      .queue[Transaction](
+        bufferSize = 1024,
+        capacityCounter = metrics.daml.SoX.conflictQueueCapacity,
+        lengthCounter = metrics.daml.SoX.conflictQueueLength,
+        delayTimer = metrics.daml.SoX.conflictQueueDelay,
+      )
+      .map { indexService.cacheOffset() -> _ }
```

Comment:
This line can be moved after the `mapAsync` below it, and thus allow changing that `mapAsync` to `mapAsyncUnordered`

Reply:
Ah, you're right 👍 Good catch